db/Pos/DB/Pure.hs

```diff
@@ -0,0 +1,127 @@
+{-# LANGUAGE RankNTypes          #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TemplateHaskell     #-}
+
+-- | Pure database implementation using 'Data.Map'. More efficient
+-- option would be something akin to muteble prefix
+-- tree. http://hackage.haskell.org/package/bytestring-trie might do
+-- better though it's immutable.
+
+module Pos.DB.Pure
+       ( DBPureMap
+       , DBPure
+       , pureBlockIndexDB
+       , pureGStateDB
+       , pureLrcDB
+       , pureMiscDB
+       , pureBlocksStorage
+
+       , MonadPureDB
+
+       , DBPureVar
+       , newDBPureVar
+
+       , dbGetPureDefault
+       , dbIterSourcePureDefault
+       , dbPutPureDefault
+       , dbDeletePureDefault
+       , dbWriteBatchPureDefault
+       ) where
+
+import           Universum
+
+import           Control.Lens                 (at, makeLenses)
+import           Control.Monad.Trans.Control  (MonadBaseControl)
+import           Control.Monad.Trans.Resource (ResourceT)
+import qualified Data.ByteString              as BS
+import           Data.Conduit                 (Source)
+import qualified Data.Conduit.List            as CL
+import           Data.Default                 (Default (..))
+import qualified Data.Map                     as M
+import qualified Database.RocksDB             as Rocks
+import           Ether.Internal               (HasLens (..))
+
+import           Pos.Binary.Class             (Bi)
+import           Pos.Core                     (HeaderHash)
+import           Pos.DB.Class                 (DBIteratorClass (..), DBTag (..), IterType,
+                                               iterKeyPrefix)
+import           Pos.DB.Functions             (processIterEntry)
+import           Pos.Util.Concurrent.RWVar    as RWV
+
+-- | Bytestring to Bytestring mapping mimicking rocks kv storage.
+type DBPureMap = Map ByteString ByteString
+
+-- | Pure database datatype. Includes 4 subdatabases mocking rocks dbs
+-- and a map for storing blocks data (we store them in files directly
+-- in real implementation).
+data DBPure = DBPure
+    { _pureBlockIndexDB  :: DBPureMap
+    , _pureGStateDB      :: DBPureMap
+    , _pureLrcDB         :: DBPureMap
+    , _pureMiscDB        :: DBPureMap
+    , _pureBlocksStorage :: Map HeaderHash ByteString
+    }
+
+makeLenses ''DBPure
+
+instance Default DBPure where
+    def = DBPure mempty mempty mempty mempty mempty
+
+type DBPureVar = RWV.RWVar DBPure
```

Comment:
Why isn't `IORef` enough here? I think the best practice is: if `IORef` is enough, use `IORef`. FYI: there is `atomicModifyIORef`, etc.
Another good practice is to use more widely used primitives, rather than our own ad-hoc modules whenever possible.